### ISSUE TITLE
Problem regarding implementation of leaf nodes

### ISSUE THREAD

**Issue by quantumiracle** [AaronX121/Soft-Decision-Tree#4]
Hi,

Thanks for your open-source implementation. However, I notice one potential problem in your implementation that confused me:

I notice that in `forward()`, `_mu` which is the path probability of each node, is passed in to `self.leaf_nodes()` as input. It means the path probability is multiplied by the weights of `self.leaf_nodes()` layer, while the softmax is taken later on in the cross-entropy loss criterion. However, as in the paper, we are expecting an average over distributions from all leaf nodes (as opposed to the maximized-path probability), so would it be more accurate to take the softmax first for each leaf node logits and then multiplied by the path probability?

Best,
Zihan

**Comment by xuyxu** [comment]
What you are referring to cannot be achieved as long as we use nn.Linear module to simulate leaf nodes as a whole. Take a look at modified codes in `SDT.py`  presented in the figure below, where parameters on leaf nodes are isolated. Also, remember to **change the criterion in main.py (Line:33) to nn.NLLLoss()**. I think this modified version solves your problem.

I conducted one sanity check on this modified version using MNIST/Fashion-MNIST, and it turned out there was no perceivable improvement. Therefore, I decided to keep current implementation unchanged for the sake of acceleration.

Please let me know if this modified version has any problem, or it somehow is superior to current implementation. Looking forward to your response.

![tmp](https://user-images.githubusercontent.com/22359569/81036218-d8cee180-8ed0-11ea-86cf-4aa819c1102f.png)

**Comment by quantumiracle** [comment]
Thanks for your reply. 

Actually I also tried similar modification yesterday, and it also doesn't seem to provide significant advantages, i.e. ~92% accuracy with maximized path probability, ~93.8% accuracy with average path probability. It's just odd to take average on logits then softmax later.

**Comment by blindFS** [comment]
@xuyxu @quantumiracle Excuse me, guys, I have a question about model inferencing.
![image](https://user-images.githubusercontent.com/568457/109285504-22bbf600-785c-11eb-856e-cc1e79427d6d.png)
According to current implementation, the output at inference time is the average of of all leaf nodes.
I wonder how much worse will it perform if predicting from the single leaf node with the largest path prob at evaluation.

Also, I'm interested in a tree model with minimal inference cost, and finding the most probable leaf still requires all the computation overhead. Will the accuracy be badly hampered if it chooses the left/right direction greedily at each inner node at inference time? Curious about your opinions.

**Comment by xuyxu** [comment]
Hi @blindFS, thanks for your interests.

For the first problem, if I remember the result correctly, the performance of only using the leaf node with the largest path probability is clearly worse than using the average predictions from all leaf nodes.

**EDIT**: My thoughts on why the performance is worse: The model complexity of soft decision tree that outputs the prediction from the leaf node with the largest path probability is much smaller than that one using the weighted predictions from all leaf nodes.

For the second problem, you may have interests in this paper, which proposes a novel activation function for soft decision tree that makes conditional computation (i.e., no need to traverse through all nodes) possible: http://proceedings.mlr.press/v119/hazimeh20a/hazimeh20a.pdf

**Comment by quantumiracle** [comment]
Hi,
I agree with @xuyxu 's results on the first question. 
For the second one, I tested the greedy choice of direction on each inner node (but also truncate multivariate node into univariate node ), for the case using SDT as policy function approximators in RL. In general case, the performance drop is significant, see in my paper (https://arxiv.org/abs/2011.07553).

**Comment by blindFS** [comment]
Thank u both for such valuable info! That could save me lots of effort.
